### ISSUE TITLE
Move a clip in ripple editing mode.

### ISSUE THREAD

**Issue by cianoc** [RustyDAW/gui-explorations#5]
## Summary
With ripple editing when I make a change to a clip, all clips to the right of it will move so as to preserve existing spacing.

## Frequency
3

## Importance
3

## Which Users would Use this
Ordinary and Power Users

## Steps in this User Story
1. Switch on Ripple Editing Mode
2. Drag a clip to another place on the timeline
3. Clips to the right of the clip on the track should remain where they were while the clip is dragged to it's new position
    + A marker on the timeline should show where the clip was on the timeline (and where the clip will return to if the user abandons the 'move' operation)
4. When the clip is released in it's new position - the following should happen:
    + **Clip's Original Position:** - Clips to the right of this position should move left (by the size of the clip) to occupy the space that it just vacated.
    + **Clip's New Position** - Clips at, or to the right of, this position should move right (by the size of the clip).
5. DAW remains in Rippled Editing Mode

## Questions
1. What happens if I place a clip over an existing clip?
   + Does that clip move right (along with all the clips to the right of it).
   + Does the clip split at the point the new clip is inserted?
   + Maybe it's a crossfade?
   + Maybe it's impossible to drop a clip on an existing clip?
   + Maybe we ask the user what they want to do in a modal?
2. What happens with crossfades
   + Can I move a clip into the next one in order to do a crossfade as this seems to contradict the premise of a crossfade.
   + Maybe crossfades should be another(!) mode.

## Acceptance Criteria
TBD

## Benchmarks
- Reaper?

## Additional Notes
Anything else that would support this use case should go here. Pictures are nice.

**Comment by WeirdConstructor** [comment]
> ## Questions
> 1. What happens if I place a clip over an existing clip?

Most intuitive for me would be:
- Insert clip before or after the dropped on clip depending on how close the mouse pointer is to the corresponding edge.
- If there is not enough space for the inserted clip:
    - The idea is: The user probably wants to align the clip with the clip they dropped on.
    - If inserted on the left with not enough space: Overlap with the preceding clip
    - If inserted on the right with not enough space: Overlap with the following clip
- Do not do ripple editing unless the user explicitly enabled this mode.
If you edit a piece of music, the timings of the clips in one track usually are synchronized with
the clips on other tracks. Imagine the user dropping the clip at the start of the song
and suddenly the whole melody is off beat.

**Comment by ollpu** [comment]
> If you edit a piece of music, the timings of the clips in one track usually are synchronized with
the clips on other tracks. Imagine the user dropping the clip at the start of the song
and suddenly the whole melody is off beat.

This applies to a lot of use cases, which is why there's often also a mode for global ripple editing that applies to other tracks as well.

A global ripple edit encompasses a lot more edge cases:
 - What happens to parallel clips (that don't necessarily line up with the clip(s) being edited)?
 - Being global, it also makes sense to affect global timeline properties like tempo changes - meaning clips stay tied to the tempo changes happening in their respective musical time.
    - **Not like this:** Adobe Premiere has a flexible global ripple edit mode, and it works with most tools on the timeline. However, it doesn't affect global keyframes, such as ones used for track volume at all. This makes the features mutually exclusive in a workflow.

**Comment by cianoc** [comment]
> Do not do ripple editing unless the user explicitly enabled this mode.
Yes I think it's stated in the user story that ripple editing should be enabled and is not the default.

Does it need to be made more explicit? I'm thinking we probably need a story for ripple mode itself.

**Comment by cianoc** [comment]
@ollpu - the complexity in your comment makes me think we may need two sets of stories:
1. Global Ripple Mode
2. Track Ripple Mode

**Comment by ollpu** [comment]
@cianoc Agreed, they are quite different.
